File: sorting/Python/sort_an_array_quick.py
class Solution:
    def sortArray(self, nums: list[int]) -> list[int]:
        self.quicksort(nums, 0, len(nums)-1)
        return nums

    def quicksort(self, arr, s, e):
        pivot = arr[e]
        left = s
        if not (e-s+1) > 1:
            return arr
        for i in range(s,e):
            if arr[i] < pivot:
                tmp = arr[left]
                arr[left] = arr[i]
                arr[i] = tmp
                left += 1
        arr[e] = arr[left]
        arr[left] = pivot

        self.quicksort(arr, s, left-1)
        self.quicksort(arr, left+1, e)

        return arr

File: sorting/C_Cpp/sort_an_array_quick.c
int* quicksort(int* arr, int s, int e);


int* sortArray(int* nums, int numsSize, int* returnSize) {
    int n = numsSize;
   *returnSize = n;
    quicksort(nums, 0, n-1);
    return nums;
}

int* quicksort(int* arr, int s, int e){
    int left = s;
    int pivot = arr[e];

    if ((e-s+1) <= 1){
        return arr;
    }

    for (int i=s; i<e; i++) {
        int tmp;
        if (arr[i] < pivot) {
            tmp = arr[left];
            arr[left] = arr[i];
            arr[i] = tmp;
            left++;
        }
    }

    arr[e] = arr[left];
    arr[left] = pivot;
    quicksort(arr, s, left-1);
    quicksort(arr, left+1, e);
    return arr;
}
